dockstore-webservice/src/main/java/io/dockstore/webservice/resources/DockerRepoResource.java

```diff
@@ -184,12 +180,43 @@ public DockerRepoResource(final HttpClient client, final SessionFactory sessionF
             LOG.info("Grabbing " + registry.getFriendlyName() + " repos");
 
             updatedTools.addAll(abstractImageRegistry
-                .refreshTools(userId, userDAO, toolDAO, tagDAO, fileDAO, fileFormatDAO, client, githubToken, bitbucketToken, gitlabToken,
-                    organization, eventDAO, dashboardPrefix));
+                .refreshTools(userId, userDAO, toolDAO, tagDAO, fileDAO, fileFormatDAO, client, githubToken, bitbucketToken,
+                    gitlabToken, organization, eventDAO, dashboardPrefix));
         }
         return updatedTools;
     }
 
+    List<Tool> refreshToolsForUser(Long userId, String organization, String repository) {
+        refreshBitbucketToken(userId);
+
+        // Get user's quay and git tokens
+        List<Token> tokens = tokenDAO.findByUserId(userId);
+        Token quayToken = Token.extractToken(tokens, TokenType.QUAY_IO);
+        Token githubToken = Token.extractToken(tokens, TokenType.GITHUB_COM);
+        Token bitbucketToken = Token.extractToken(tokens, TokenType.BITBUCKET_ORG);
+        Token gitlabToken = Token.extractToken(tokens, TokenType.GITLAB_COM);
+
+        // with Docker Hub support it is now possible that there is no quayToken
+        checkTokens(quayToken, githubToken, bitbucketToken, gitlabToken);
+
+        // Get a list of all namespaces from Quay.io only
+        if (quayToken == null) {
+            throw new CustomWebApplicationException("Missing required Quay.io token", HttpStatus.SC_BAD_REQUEST);
+        }
+        QuayImageRegistry registry = new QuayImageRegistry(quayToken);
+        return new ArrayList<>(
```

Comment:
Why do you wrap it with an ArrayList and not just return `registry.refreshTool(...` directly?

Reply:
endpoint uses ArrayList